Add manifest write functions and tests for the functions

matthewrsj opened the pull request:
Signed-off-by: Matthew Johnson <matthew.johnson@intel.com>

matthewrsj committed:
Add manifest write functions and tests for the functions

Signed-off-by: Matthew Johnson <matthew.johnson@intel.com>

bryteise commented on src/swupd/manifest.go:
```diff
@@ -226,3 +226,111 @@ func (m *Manifest) ReadManifestFromFile(f string) error {
+		return err
+	}
+
+	// return error code or nil from w.Flush()
```
Does the error from w.Flush() get overwritten if there is an error in the defer function?

matthewrsj replied:
oh perhaps. Maybe I should do:
```
err = w.Flush()
return err
```

matthewrsj replied:
done

icarus-sparry commented on src/swupd/manifest.go:
```diff
@@ -226,3 +226,113 @@ func (m *Manifest) ReadManifestFromFile(f string) error {
+}
+
+// writeManifestFileHeader writes the header of a manifest to the file
+func writeManifestFileHeader(m *Manifest, w *bufio.Writer) error {
```
Should be able to just declare w as *io.Writer, no need to require it to be a *bufio.Writer.

cmarcelo replied:
More precisely as: w io.Writer (without the pointer).

This function is making assumptions about how w handles errors (which hold true for *bufio.Writer but not for an arbitrary io.Writer), so you'll have to do more changes to handle errors for writes.

To make this less noisy, you can merge the prints into a single one:

    _, err := fmt.Fprintf(w, `MANIFEST\t%v
    version:\t%v
    ...`, m.Header.Format, ...)

icarus-sparry commented on src/swupd/manifest.go:
```diff
@@ -226,3 +226,113 @@ func (m *Manifest) ReadManifestFromFile(f string) error {
+		return err
+	}
+
+	if _, err = w.WriteString(fmt.Sprintf("MANIFEST\t%v\n", m.Header.Format)); err != nil {
```
Worth looking at  [errors are values](https://blog.golang.org/errors-are-values) to see if the idea of errorwriter might help. Of course this sort of negates my comment about io.Writer vs bufio.Writer as bufio.Writer is an errorwriter

matthewrsj replied:
Interesting, it looks like this means errors in the previous writes will be caught when `w.Flush()` is called later?

icarus-sparry commented on src/swupd/manifest.go:
```diff
@@ -226,3 +226,113 @@ func (m *Manifest) ReadManifestFromFile(f string) error {
+
+	line := fmt.Sprintf("%v\t%v\t%v\t%v\n",
+		flags,
+		file.getHashString(),
```
At least my intention here was that you could simply say file.Hash here, and you shouldn't need the getHashString method at all, the String method on hashval should take care of it.

In particular I could imagine in the future storing the hash as a 32 byte binary object, rather than a 64 (+1 for terminating NUL) string of hex characters to reduce space.

icarus-sparry commented on src/swupd/manifest.go:
```diff
@@ -226,3 +226,113 @@ func (m *Manifest) ReadManifestFromFile(f string) error {
+}
+
+// writeManifestFileEntries writes all the entries in files to the file
+func writeManifestFileEntries(m *Manifest, w *bufio.Writer) error {
```
Not sure this is worth a separate function. If it is then it should take an *io.Writer.

icarus-sparry commented on src/swupd/manifest_test.go:
```diff
@@ -254,3 +258,90 @@ func TestReadManifestFromFileEmpty(t *testing.T) {
+	if err != nil {
+		return false, err
+	}
+
```
Missing os.Close, preferably deferred to catch all the exception paths.

icarus-sparry commented on src/swupd/manifest_test.go:
```diff
@@ -254,3 +258,90 @@ func TestReadManifestFromFileEmpty(t *testing.T) {
 }
+
+func compareFiles(file1path string, file2path string) (bool, error) {
+	chunkSize := 64000
```
Before you start reading, do a couple of os.Stat calls. If the sizes differ then you don't need to do all the I/O to read them.

64000 is a reasonable number, but might be worth switching to 65536 in the vague hope that you might get some zero copying advantages from the kernel

icarus-sparry commented on src/swupd/manifest_test.go:
```diff
@@ -254,3 +258,90 @@ func TestReadManifestFromFileEmpty(t *testing.T) {
+
+	for {
+		b1 := make([]byte, chunkSize)
+		_, err1 := f1.Read(b1)
```
Throwing away the number of bytes read is just wrong. Of course you probably don't want to invent co-routines, so you get into issues where read could return a less than full buffer... Probably best to avoid being clever here and just use bufio.

matthewrsj commented:
@icarus-sparry pushed some updates.

matthewrsj committed:
Remove unnecessary getHashString function

The getHashString function is not necessary with the hash.String()
function implementation.

Signed-off-by: Matthew Johnson <matthew.johnson@intel.com>

matthewrsj commented:
I'm going to merge this soon unless anyone objects. I need the manifest writers at this point.

matthewrsj committed:
Update header validation to use latest

CheckHeaderPopulated changed to CheckHeaderIsValid

Signed-off-by: Matthew Johnson <matthew.johnson@intel.com>

cmarcelo commented on src/swupd/manifest.go:
```diff
@@ -242,3 +242,85 @@ func (m *Manifest) ReadManifestFromFile(f string) error {
 }
+
+// getManifestTimeStamp returns the unix timestamp of m.Header.TimeStamp
+func (m *Manifest) getManifestTimeStamp() int64 {
```
Minor: since this is a method on the manifest, it is implied that it refers to the manifest, so you could call this just getTimeStamp() or even timeStamp(). See also: https://golang.org/doc/effective_go.html#Getters

cmarcelo replied:
Minor: or just drop this function entirely.

matthewrsj replied:
yeah it can be removed

cmarcelo commented on src/swupd/manifest.go:
```diff
@@ -242,3 +242,85 @@ func (m *Manifest) ReadManifestFromFile(f string) error {
+
+	// bufio.Writer is an errWriter, so errors will be returned when the calling
+	// function calls w.Flush()
+	w.WriteString(fmt.Sprintf("MANIFEST\t%v\n", m.Header.Format))
```
Minor: use fmt.Fprintf.

cmarcelo commented on src/swupd/manifest_test.go:
```diff
@@ -257,3 +262,117 @@ func TestInvalidManifests(t *testing.T) {
+	now := time.Now()
+	m := Manifest{Header: ManifestHeader{TimeStamp: now}}
+	if m.getManifestTimeStamp() != now.Unix() {
+		t.Error("gerManifestTimestamp returned unexpected result")
```
Minor: Typo "ger".

cmarcelo commented on src/swupd/manifest_test.go:
```diff
@@ -257,3 +262,117 @@ func TestInvalidManifests(t *testing.T) {
+
+func TestWriteManifestFileBadHeader(t *testing.T) {
+	m := Manifest{Header: ManifestHeader{}}
+	path := "testdata/manifest.result"
```
Minor: use ioutil.TempFile to create a temporary file (you can grab the Name() from the returned file). That way we don't mix testdata input with temporary stuff and we don't need to care about tests using different temporary names.

cmarcelo commented on src/swupd/manifest_test.go:
```diff
@@ -257,3 +262,117 @@ func TestInvalidManifests(t *testing.T) {
+		t.Fatal("ReadManifestFromFile did not add file entried to the file list")
+	}
+
+	newpath := "testdata/manifest.result"
```
Minor: see suggestion about using TempFile.

cmarcelo reviewed (approved):
I was late to the party, so consider all my comments "Minor" and things that can be handled later on.

matthewrsj committed:
Make AllZeroHash addressable instead of a const

Signed-off-by: Matthew Johnson <matthew.johnson@intel.com>

matthewrsj commented:
@cmarcelo I fixed up the typo and tempfile comments. The other things I'll leave until later.
